### PR TITLE
dirs,overlord/snapstate: add Soft and Hard refresh checks

### PR DESCRIPTION
**Pull request by zyga**
As a part of "refresh app awareness" feature snapd should be able to
check if a snap is busy just before starting the refresh process and
then re-check just before stopping services and performing link changes.

There are two checks defined as a part of this flow: a soft check
that filters out processes that belong to services and a hard check that
simply looks if there are any processes running (that associate with given
snap).

This patch introduces this logic without actually using it yet. The soft
refresh check takes a snap.Info to know about apps and hooks. It looks
at the freezer cgroup which tracks all processes in a given snap
(regardless of origin) and then looks at the per-service PIDs cgroup
that tracks per-service processes. If there are any non-service processes
around then a snap is deemed busy.

Supplemental logic inspects apps and hooks to provide a descriptive
error message. Even though the process is partially racy and can provide
false positives or false negatives (e.g. a snap being busy just stopped
being busy or a snap not being busy may have just started to be busy)
this is deliberate. There is no freezing or locking at this step.

The second check is a hard check. After stopping services we can run the
hard check that simply looks if there are any processes left.

This time there is no synchronization internally but externally
(in the update logic) snapd will stop services, grab a lock and only then
perform the hard check and unlink a snap. This will ensure that the
process is free of any races. The exact semantics of the second lock
is not defined yet but it is explicitly orthogonal of this code.

The lock may be either the existing per-snap lock or a new lock that
has no timeout associated with it. This second lock would stall
startup of any new snap apps or hooks until the refresh process
can link the new revision.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: overlord/snapstate/refresh.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+// services are running, that is, that no non-service apps or hooks are
+// currently running.
+//
+// Since services are stopped during the update this provides a good early
+// precondition check.  The check is also deliberately racy, both at the level
+// of processes forking and exiting and at the level of snap-confine launching
+// new commands. The hard check needs to be synchronized but the soft check
+// doesn't require this since it would serve no purpose. After the soft check
+// passes the user is free to start snap applications and block the hard check.
+func SoftRefreshCheck(info *snap.Info) error {
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: wondering if a longer name "SoftNothingRunningRefreshCheck" would be better

File: overlord/snapstate/refresh.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+//
+// The check is designed to run late in the refresh pipeline, after stopping
+// snap services. At this point services should be stopped, hooks should no
+// longer run, and applications should be barred from running externally (e.g.
+// by grabbing the per-snap lock around that phase of the update).
+//
+// The check looks at the set of PIDs in the freezer cgroup associated with a
+// given snap. Presence of any processes indicates that a snap is busy and
+// refresh cannot proceed.
+func HardRefreshCheck(snapName string) error {
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: same question

File: overlord/snapstate/refresh.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		sort.Ints(pids)
+		return &BusySnapError{pids: pids, snapName: snapName}
+	}
+	return nil
+}
+
+// BusySnapError indicates that snap has apps or hooks running and cannot refresh.
+type BusySnapError struct {
+	pids          []int
+	snapName      string
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: I think snapName should be the first field

File: overlord/snapstate/refresh.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+// external synchronization system applied (e.g. all processes are frozen) at
+// the time the snapshot was taken.
+//
+// The list is intended for snapd to forcefully kill all processes for a forced
+// refresh scenario.
+func (err BusySnapError) Pids() []int {
+	return err.pids
+}
+
+// parsePid parses a string as a process identifier.
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: I would move the parse* helpers at the end of the file

File: overlord/snapstate/refresh.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+//
+// The check looks at the set of PIDs in the freezer cgroup associated with a
+// given snap. Presence of any processes indicates that a snap is busy and
+// refresh cannot proceed.
+func HardRefreshCheck(snapName string) error {
+	pidSet, err := pidSetOfSnap(snapName)
+	if err != nil {
+		return err
+	}
+	if len(pidSet) > 0 {
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: even for the hard check we should ignore services that have refresh-mode: endure,  the snap is asking us explicitly to let them run, not much we can do

Comment (zyga) [reply overlord/snapstate/refresh.go]: Ah, this is interesting. So the hard check should still ignore services with endure mode? I guess we don't have many of those but this still means that the original problem remains for a subset of our users.

**Comment by zyga** [comment]
I will return here tomorrow.

**Comment by zyga** [comment]
I am working on handling endure mode services now. This will be ready later today.

File: overlord/snapstate/refresh.go (hunk 5 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: Endure mode services are now handled.

File: overlord/snapstate/refresh.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+type softRefreshCheckDelegate struct{}
+
+func (*softRefreshCheckDelegate) canAppRunDuringRefresh(app *snap.AppInfo) bool {
+	return app.IsService()
+}
+
+func (*softRefreshCheckDelegate) canHookRunDuringRefresh(hook *snap.HookInfo) bool {
+	return false
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: is this going to be true in some cases?

not immediately interesting but hooks are interesting because we run them, not the user, and they have a fixed timeout (which is quite high though :/ ) so there would be some logic into retrying for them (that add complexity though)

File: overlord/snapstate/refresh.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+type hardRefreshCheckDelegate struct{}
+
+func (*hardRefreshCheckDelegate) canAppRunDuringRefresh(app *snap.AppInfo) bool {
+	// TODO: use a constant instead of "endure"
+	return app.IsService() && app.RefreshMode == "endure"
+}
+
+func (*hardRefreshCheckDelegate) canHookRunDuringRefresh(hook *snap.HookInfo) bool {
+	return false
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: is this going be true in some cases?

File: overlord/snapstate/refresh.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaplock"
+)
+
+type refreshCheckDelegate interface {
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: not a fan of Delegate here,   is one of those generic "pattern name", maybe refreshExpecationsChecker ? anyway see questions  below, it's unclear we really need both methods

File: snap/snaplock/lock.go [hunk 9]
```diff
@@ -17,7 +17,7 @@
  *
  */
 
-package mount
+package snaplock
```
Comment (pedronis) [inline snap/snaplock/lock.go]: this could be have been its own PR,  it's  a bit unclear if it belongs under snap (not that I we have a clear better place) ... is this lock intrinsic to snaps as a format and how they are meant to be delivered?  (the line is blurry but most of the content of snap/ follows that line)

File: overlord/snapstate/refresh.go (hunk 6 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: For now it's always going to be false. If you want I can simplify this.

File: overlord/snapstate/refresh.go (hunk 8 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: I can probably fold the hook method into the generic check and just use a single callback.

File: snap/snaplock/lock.go (hunk 9 above)
Comment (zyga) [reply snap/snaplock/lock.go]: I could document the meaning of the lock. Not sure if you have a concrete proposal on how to move forward.

File: overlord/snapstate/refresh.go (hunk 7 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: Not that I can predict now.

File: overlord/snapstate/refresh.go (hunk 8 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: Done now.

File: snap/snaplock/lock.go (hunk 9 above)
Comment (pedronis) [reply snap/snaplock/lock.go]: given is used by snap-update-ns  and is the same lock as snap-confine,  I would move it to cmd/snaplock for now (it's not ideal but we have some other things we use out of cmd that are a bit odd)

Comment (zyga) [reply snap/snaplock/lock.go]: Hmm, I find that very odd, cmd has no non-command packages. Why is that better in your eyes than `snap/snaplock`? As we work on this feature I expect we will need another type of lock that would also make sense to add to `snaplock` (the one with inhibit semantics that can be held longer).

Comment (pedronis) [reply snap/snaplock/lock.go]: we already have cmd/snapinfo.go  with he same kind of problem (we need a place for that and this but for now they can go there),  as I said I really prefer to maintain  some distinction between more intrinsic aspect of snaps,  and more implementation details for them

Comment (zyga) [reply snap/snaplock/lock.go]: Are you proposing that `snaplock` then moves to `cmd/snaplock.go`?

Comment (pedronis) [reply snap/snaplock/lock.go]: no, I'm still asking for a cmd/snaplock package

Comment (pedronis) [reply snap/snaplock/lock.go]: if it sounds too much like a non-existing command, snaplocking is also a fine name

Comment (pedronis) [reply snap/snaplock/lock.go]: Also it doesn't seem to have a comment about the fact that this kind of lock is created also from snap-confine, a pointer to the corresponding c code.

File: overlord/snapstate/refresh.go [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+// HardNothingRunningRefreshCheck looks if there are any processes alive.
+//
+// The check is designed to run late in the refresh pipeline, after stopping
+// snap services. At this point services should be stopped, hooks should no
+// longer run, and applications should be barred from running externally (e.g.
+// by grabbing the per-snap lock around that phase of the update).
+//
+// The check looks at the set of PIDs in the freezer cgroup associated with a
```
Comment (pedronis) [inline overlord/snapstate/refresh.go]: the comment about the freezer group seems untrue, we are only looking at pid groups, no?

Comment (zyga) [reply overlord/snapstate/refresh.go]: Yes, the comment needs to be updated now. Thank you!

**Comment by zyga** [comment]
I realised that the main commit message is a little bit out of sync. I will correct it tomorrow.

File: overlord/snapstate/refresh.go [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+	// may fork or exit, we will have per-security-tag information about
+	// what is running.
+	lock, err := snaplock.OpenLock(info.SnapName())
+	if err != nil {
+		return err
+	}
+	defer lock.Close()
+	if err := lock.Lock(); err != nil {
+		return err
+	}
```
Comment (mvo5) [inline overlord/snapstate/refresh.go]: Do we need a "defer lock.Unlock()" here or is the "lock.Close" above sufficient? If the later, maybe a comment because AIUI its critical that we unlock or the snap-app won't be able to run, yes?

File: overlord/snapstate/refresh.go [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+	var busyAppNames []string
+	var busyHookNames []string
+	var busyPIDs []int
+
+	canHookRunDuringRefresh := func(hook *snap.HookInfo) bool {
```
Comment (mvo5) [inline overlord/snapstate/refresh.go]: I think this needs a TODO comment at least with the plan what will happen here. Or alternative comment the code below that calls "canHookRunDuringRefresh" and add a TODO there (might be cleaner this way).

File: overlord/snapstate/refresh.go [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+			return err
+		}
+		if len(PIDs) > 0 {
+			busyAppNames = append(busyAppNames, name)
+			busyPIDs = append(busyPIDs, PIDs...)
+		}
+	}
+
+	for name, hook := range info.Hooks {
+		if canHookRunDuringRefresh(hook) {
```
Comment (mvo5) [inline overlord/snapstate/refresh.go]: Thinking about this - do we know already when hooks can run during refreshes? If we don't have any use-case yet, maybe this is a case of YAGNI? I mean, we can always add it if we need it.

File: overlord/snapstate/refresh.go [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+// The check is designed to run late in the refresh pipeline, after stopping
+// snap services. At this point non-enduring services should be stopped, hooks
+// should no longer run, and applications should be barred from running
+// externally (e.g. by using a new inhibition mechanism for snap run).
+//
+// The check fails if any process belonging to the snap, apart from services
+// that are enduring refresh, is still alive. If a snap is busy it cannot be
+// refreshed and the refresh process is aborted.
+func HardNothingRunningRefreshCheck(info *snap.Info) error {
+	return genericRefreshCheck(info, func(app *snap.AppInfo) bool {
```
Comment (mvo5) [inline overlord/snapstate/refresh.go]: The difference between Hard/Soft check is just "RefreshMode == "ensure", yes? This makes me wonder if instead of this indirection and two different exported functions and one generic helper we could simply have a single function and a flag, something like (names are just strawmans):
```
type RefreshCheckFlags {
    HardNothingRunning bool // HardNothingRunning looks if there are any undesired processes alive.
}
func RefreshCheck(info *snap.Info, flags RefreshCheckFlags) error {
...
```
or do we expect that canAppRunDuringRefresh grow soon? If not, maybe YAGNI? But I understand this is partly personal preference so happy to be ignored if you and @pedronis  think this does not really help making things easier to follow.

File: overlord/snapstate/refresh.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+	snapName      string
+	pids          []int
+	busyAppNames  []string
+	busyHookNames []string
+}
+
+// Error formats an error string describing what is running.
+func (err *BusySnapError) Error() string {
+	switch {
+	case len(err.busyAppNames) > 0 && len(err.busyHookNames) > 0:
```
Comment (mvo5) [inline overlord/snapstate/refresh.go]: Nice one error

File: overlord/snapstate/refresh.go (hunk 11 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: The close is sufficient. This is a file-descriptor based lock so closing it is totally enough. I can add a comment.

File: overlord/snapstate/refresh.go (hunk 12 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: It's just for consistency with `canAppRunDuringRefresh`. There's nothing to do here for now. I can add a comment.

File: overlord/snapstate/refresh.go (hunk 13 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: Currently hooks cannot run during refreshes. This form is just for consistency with app checks but we must check that no hooks are running.

File: overlord/snapstate/refresh.go (hunk 14 above)
Comment (zyga) [reply overlord/snapstate/refresh.go]: I don't know how this will behave in the future. I would like to keep the two public functions explicitly because it matches our discussions from Cape Town and earlier. The soft and hard checks will still exist even though their details may vary. I don't think this warrants a merging (given that we have the private generic function already).

Comment (pedronis) [reply overlord/snapstate/refresh.go]: Yes, agreed, I think  the two functions map the way we discussed about this.

**Comment by zyga** [comment]
@jdstrand I'm merging this to make progress. The code is currently unused and the upcoming branch that enables it will only do so behind a feature flag. I am happy to incorporate any changes requested from a post-merge review.

File: overlord/snapstate/refresh.go [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+func (err BusySnapError) Pids() []int {
+	return err.pids
+}
+
+// parsePid parses a string as a process identifier.
+func parsePid(text string) (int, error) {
+	pid, err := strconv.Atoi(text)
+	if err == nil && pid <= 0 {
+		return 0, fmt.Errorf("cannot parse pid %q", text)
+	}
```
Comment (jdstrand) [inline overlord/snapstate/refresh.go]: It would be highly unusual if pid == 1 too and I'm a little concerned if there is a bug that we might forcibly kill pid 1 and nuke the system. That said, considering snapd running in containers, pid 1 might not be as interesting in the container. I bring this up not to change anything but to simply think through if this is something we'd want to explicitly avoid here.
